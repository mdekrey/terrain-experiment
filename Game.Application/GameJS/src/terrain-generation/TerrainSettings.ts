import {
  DataDrivenConstructor,
  DataDrivenInput
} from "../utils/DataDrivenComposition";
import { libnoise } from "libnoise";
import { VisualTerrainType } from "./VisualTerrainType";
import { dataDrivenTerrainOnly, TerrainSpec } from "./terrain-specifications";

type TerrainSpecCreation = DataDrivenConstructor<number, libnoise.ModuleBase>;
type TerrainSpecDto = DataDrivenInput<any, TerrainSpec<VisualTerrainType>>;
export interface TerrainSettingsDto {
  tempsStep: number[];
  humidityStep: number[];
  altitudeStep: number[];
  humidityCurve: { slope: number; offset: number };
  temperaturePenalty: { slope: number; offset: number };
  humidity: TerrainSpecCreation;
  heat: TerrainSpecCreation;
  altitude: TerrainSpecCreation;
  feature: TerrainSpecCreation;
  caveIndicator: TerrainSpecCreation;
  caveSeeds: TerrainSpecCreation;
  visualizationSpec: TerrainSpecDto;
  detailVisualizationSpec: TerrainSpecDto;
}

export interface TerrainSettings {
  tempsStep: number[];
  humidityStep: number[];
  altitudeStep: number[];
  humidityCurve: { slope: number; offset: number };
  temperaturePenalty: { slope: number; offset: number };
  humidity: TerrainSpecCreation;
  heat: TerrainSpecCreation;
  altitude: TerrainSpecCreation;
  feature: TerrainSpecCreation;
  caveIndicator: TerrainSpecCreation;
  caveSeeds: TerrainSpecCreation;
  visualizationSpec: TerrainSpec<VisualTerrainType>;
  detailVisualizationSpec: TerrainSpec<VisualTerrainType>;
}

export function terrainSettingsFromDto({
  visualizationSpec,
  detailVisualizationSpec,
  ...rest
}: TerrainSettingsDto) {
  return {
    ...rest,
    visualizationSpec: dataDrivenTerrainOnly(visualizationSpec),
    detailVisualizationSpec: dataDrivenTerrainOnly(detailVisualizationSpec)
  };
}

export function humidityCurve(
  humidityCurve: TerrainSettings["humidityCurve"],
  originalHumidity: number,
  heat: number
) {
  return (humidityCurve.offset + heat * humidityCurve.slope) * originalHumidity;
}

export function temperaturePenalty(
  temperaturePenalty: TerrainSettings["temperaturePenalty"],
  altitude: number
) {
  return Math.max(
    0,
    altitude * temperaturePenalty.slope + temperaturePenalty.offset
  );
}
