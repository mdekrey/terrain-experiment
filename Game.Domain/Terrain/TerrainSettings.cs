using Game.Application.Models;
using Game.Domain.Specifications;
using LibNoise;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Game.Domain.Terrain
{
    public class TerrainSettings
    {
        public const float overworldGridSize = 0.25e-2f;
        public const float localGridSizeDiff = 1e-2f;
        public const float localGridSize = 0.25e-4f;

        public Dictionary<TemperatureCategory, float> TemperatureStep { get; set; }
        public Dictionary<HumidityCategory, float> HumidityStep { get; set; }
        public Dictionary<AltitudeCategory, float> AltitudeStep { get; set; }
        public LinearFormula HumidityCurve { get; set; }
        public LinearFormula TemperaturePenalty { get; set; }
        public ModuleBase Humidity { get; set; }
        public ModuleBase Heat { get; set; }
        public ModuleBase Altitude { get; set; }
        public ModuleBase Feature { get; set; }
        public ModuleBase CaveIndicator { get; set; }
        public ModuleBase CaveSeeds { get; set; }
        public ISpecification<ITerrainSituation, VisualTerrainType> VisualizationSpec { get; set; }
        public ISpecification<ITerrainSituation, VisualTerrainType> DetailVisualizationSpec { get; set; }

        public float CalculateTemperaturePenalty(float altitude)
        {
            return Math.Max(
              0,
              altitude * TemperaturePenalty.Slope + TemperaturePenalty.Offset
            );
        }

        public float CalculateHumidity(float originalHumidity, float heat)
        {
            return (HumidityCurve.Offset + heat * HumidityCurve.Slope) * originalHumidity;
        }

        public TerrainPoint GenerateSituation(float x, float y)
        {
            var altitude = Altitude.GetValue(x, y, 0);
            var heat = Heat.GetValue(x, y, 0) - CalculateTemperaturePenalty(altitude);
            var humidity = CalculateHumidity(Humidity.GetValue(x, y, 0), heat);
            var feature = Feature.GetValue(x, y, 0);
            var caveIndicator = CaveIndicator.GetValue(x, y, 0);
            var isCave = false;

            if (IsOnGrid(x, y))
            {
                var altitudeCategory = GetBestKey(AltitudeStep, altitude);
                if ((altitudeCategory == AltitudeCategory.None || altitudeCategory == AltitudeCategory.Hills) &&
                    caveIndicator >= 3e6 && caveIndicator <= 6e6)
                {
                    isCave = true;
                }
            }

            return new TerrainPoint(new GameCoordinate(x, y), humidity, heat, altitude, feature, isCave, this);
        }

        public static bool IsOnGrid(float x, float y)
        {

            bool IsCoordinateOnGrid(float coord) => Math.Abs(Math.Round(coord / overworldGridSize) - Math.Round(coord / localGridSize) * localGridSizeDiff) <= localGridSize;

            var onGrid = IsCoordinateOnGrid(x) && IsCoordinateOnGrid(y);

            return onGrid;
        }

        public static T GetBestKey<T>(Dictionary<T, float> steps, float target)
        {
            return (from kvp in steps
                    orderby kvp.Value
                    where kvp.Value > target
                    select kvp.Key
                    ).FirstOrDefault();
        }

    }
}
