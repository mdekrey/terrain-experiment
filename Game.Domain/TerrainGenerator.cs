using System;
using System.Collections.Generic;
using System.Text;
using LibNoise;
using LibNoise.Filter;
using LibNoise.Primitive;

namespace Game.Domain
{
    public class TerrainGenerator
    {
        private static float SumFractalRange(int octaveCount)
        {
            var pow = Math.Pow(2, octaveCount);
            return (float)((pow - 1) / pow * 2);
        }
        private static readonly int OctaveCount = 6;
        private readonly float sumFractalRange = SumFractalRange(OctaveCount);
        private readonly float ridgedRange = 2.15f;
        private readonly IModule2D humidity = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 0 }, OctaveCount = OctaveCount };
        private readonly IModule2D heat = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 1750 }, OctaveCount = OctaveCount };
        private readonly IModule2D altitude = new RidgedMultiFractal { Primitive2D = new SimplexPerlin { Seed = 3201 }, OctaveCount = OctaveCount };
        private readonly IModule2D feature = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 670 }, OctaveCount = OctaveCount, Lacunarity = 3.2f };
        private readonly IModule2D caveSeeds = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 900 }, OctaveCount = OctaveCount };

        public TerrainPoint GetTerrainAt(GameCoordinate coordinates) {
            var altitude = GetAltitude(this.altitude.GetValue(coordinates));
            var heat = GetHeat(this.heat.GetValue(coordinates), altitude);
            var humidity = GetHumidity(this.humidity.GetValue(coordinates), heat);
            return new TerrainPoint(
                coordinates,
                humidity,
                heat,
                altitude,
                (FeatureBlend(feature, coordinates) / sumFractalRange).Clamp(-1, 1)
            );
        }

        private float FeatureBlend(IModule2D feature, GameCoordinate coordinates)
        {
            const float multFactor = 6000;
            const float stepSize = 100;
            // guarantees coords as 0 <= coord < stepSize
            var onex = ((coordinates.x % stepSize) + stepSize) % stepSize;
            var oney = ((coordinates.y % stepSize) + stepSize) % stepSize;
            var altx = onex + stepSize;
            var alty = oney + stepSize;
            var xWeight = onex / stepSize;
            var yWeight = oney / stepSize;
            var invXWeight = 1 - xWeight;
            var invYWeight = 1 - yWeight;

            return feature.GetValue(onex * multFactor, oney * multFactor) * xWeight * yWeight
                + feature.GetValue(altx * multFactor, oney * multFactor) * invXWeight * yWeight
                + feature.GetValue(onex * multFactor, alty * multFactor) * xWeight * invYWeight
                + feature.GetValue(altx * multFactor, alty * multFactor) * invXWeight * invYWeight;
        }

        private float GetAltitude(float inValue) => (inValue / ridgedRange).Clamp(0, 1);

        private float GetHeat(float inValue, float altitude)
        {
            var altitudeCooling = Math.Max(0, altitude * 2 - 1.7f) * 2;
            return (inValue / sumFractalRange - altitudeCooling).Clamp(-1, 1);
        }
        private float GetHumidity(float inValue, float heat)
        {
            var dehumidifyingFactor = (0.7f + heat * 0.4f);
            return (dehumidifyingFactor  * inValue / sumFractalRange).Clamp(-1, 1);
        }
    }

    static class LibNoiseExtensions
    {
        public static float GetValue(this IModule2D module, GameCoordinate coordinates) => module.GetValue(coordinates.x, coordinates.y);

        public static float Clamp(this float target, float min, float max) => Math.Min(Math.Max(target, min), max);
    }
}
