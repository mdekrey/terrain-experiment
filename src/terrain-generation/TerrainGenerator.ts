import { Perlin } from "libnoise-ts/module/generator";
import { TemperatureCategory } from "./TemperatureCategory";
import { HumidityCategory } from "./HumidityCategory";
import { BiomeCategory } from "./BiomeCategory";
import { BiomeLabel } from "./BiomeLabel";
import { BiomeDetails } from "./BiomeDetails";

export interface TerrainResult {
    humidity: number,
    heat: number,
    altitude: number,
    temperatureCategory : TemperatureCategory,
    humidityCategory: HumidityCategory,
    biomeLabel: BiomeLabel,
    biomeCategory: BiomeCategory
}

const clamper = clamp(0, 1 - Number.EPSILON);
const toValidRange = (v: number) => {
  if (v < -1.5 || v > 1.5) {
    console.log(v);
  }
  return clamper((1.5 + v) / 3);
}

const tempsStep = [.126,.235,.406,.561,.634,.876, Number.MAX_VALUE];
const humidityCurve = (originalHumidity: number, heat: number) => (0.2 + heat * 0.8) * originalHumidity;

class PerlinAnyDirection {
  private static clamp = clamp(0, 1);
  private readonly perlin = new Perlin();
  private readonly overlap: number;

  constructor({ lacunarity = Perlin.DEFAULT_PERLIN_LACUNARITY, seed = Perlin.DEFAULT_PERLIN_SEED, overlap = 1, octaves = Perlin.DEFAULT_PERLIN_OCTAVE_COUNT }) {
    this.perlin.lacunarity = lacunarity;
    this.perlin.seed = seed;
    this.perlin.octaves = octaves;
    this.overlap = overlap;
  }

  private weight(x: number, y: number) {
    return PerlinAnyDirection.clamp(x / this.overlap) * PerlinAnyDirection.clamp(y / this.overlap);
  }

  getValue(x: number, y: number) {
    const altx = this.overlap - x, alty = this.overlap - y;
    const sets: [number,number,number][] = [
      [x, y, 0],
      [altx, y, 1],
      [x, alty, 2],
      [altx, alty, 3]
    ];
    const values = sets.filter(([x, y]) => x >= 0 && y >= 0).map(([x,y,z]) => this.perlin.getValue(x,y,z) * this.weight(x,y));
    return values.reduce((p, n) => p + n);
  }
}

export class TerrainGenerator {
  private readonly humidity = new PerlinAnyDirection({ lacunarity: 3.2, seed: 0 });
  private readonly heat = new PerlinAnyDirection({ lacunarity: 3.2, seed: 1750 });
  private readonly altitude = new PerlinAnyDirection({ lacunarity: 3.2, seed: 3200 });

  public getTerrain(x: number, y: number): TerrainResult {

    const altitude = toValidRange(this.altitude.getValue(x, y));
    const heat = toValidRange(this.heat.getValue(x, y) - Math.max(0, altitude * 2 - 1.7));
    const humidity = humidityCurve(toValidRange(this.humidity.getValue(x, y)), heat);
    const temperatureCategory = tempsStep.findIndex(v => v > heat) as TemperatureCategory;
    const humidityCategory = Math.min(
      BiomeDetails.biomeLabels[temperatureCategory].length,
      Math.floor(8 * humidity)
    ) as HumidityCategory;
    const biomeLabel =
      BiomeDetails.biomeLabels[temperatureCategory][humidityCategory];
    const biomeCategory = BiomeDetails.categoryLookup[biomeLabel];
    return {
      humidity,
      heat,
      altitude,
      temperatureCategory,
      humidityCategory,
      biomeLabel,
      biomeCategory
    };
  }
}

function clamp(min: number, max: number) {
  return (value: number) => Math.min(max, Math.max(value, min));
}
