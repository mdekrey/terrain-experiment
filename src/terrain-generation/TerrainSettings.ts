export class TerrainSettings {

    readonly tempsStep = [0.126, 0.235, 0.406, 0.561, 0.634, 0.876, Number.MAX_VALUE];

    humidityCurve(originalHumidity: number, heat: number) {
        return (0.2 + heat * 0.8) * originalHumidity;
    }

}