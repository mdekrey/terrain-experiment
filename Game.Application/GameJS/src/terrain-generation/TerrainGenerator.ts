import { TerrainSettings } from "./TerrainSettings";
import { initializePerlin, initializeRidgedMulti } from "../utils/LibNoiseUtils";
import { clamp } from "../utils/clamp";
import { TerrainPoint } from "./TerrainPoint";
import { GameCoordinates } from "../game/GameCoordinates";

const perlinRange = 1.77;
const clamper = clamp(0, 1 - Number.EPSILON);
const toValidRange = (v: number) => {
  if (process.env.NODE_ENV === "development" && Math.abs(v) > perlinRange / 2) {
    console.log(v, "not in expected perlinRange", perlinRange);
  }
  return clamper(v / perlinRange + 0.5);
};

export class TerrainGenerator {
  private readonly humidity = initializePerlin({
      lacunarity: 3.2,
      seed: 0
    });
  private readonly heat = initializePerlin({
      lacunarity: 3.2,
      seed: 1750
    });
  private readonly altitude = initializeRidgedMulti({
      lacunarity: 3.2,
      seed: 3201
    });
  private readonly feature = initializeRidgedMulti({
      lacunarity: 3.2,
      seed: 670
    });
  private readonly caveSeeds = initializeRidgedMulti({
      lacunarity: 3.2,
      seed: 900
    });
  private readonly terrainSettings: TerrainSettings;

  constructor(terrainSettings: TerrainSettings) {
    this.terrainSettings = terrainSettings;
  }

  public getTerrain(x: number, y: number): TerrainPoint {
    const altitude = toValidRange(this.altitude.getValue(x, y, 0));
    const heat = toValidRange(
      this.heat.getValue(x, y, 0) - Math.max(0, altitude * 2 - 1.7)
    );
    const humidity = this.terrainSettings.humidityCurve(
      toValidRange(this.humidity.getValue(x, y, 0)),
      heat
    );
    const feature = toValidRange(this.feature.getValue(x * 6000, y * 6000, 0));

    return new TerrainPoint(
      this.terrainSettings,
      x,
      y,
      altitude,
      heat,
      humidity,
      feature
    );
  }

  getCaveSeedAt(point: GameCoordinates) {
    return this.caveSeeds.getValue(point.x, point.y, 0) * 100000;
  }
}
