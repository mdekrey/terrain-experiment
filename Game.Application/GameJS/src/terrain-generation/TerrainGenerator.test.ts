import { TerrainGenerator } from "./TerrainGenerator";
import { TerrainSettings } from "./TerrainSettings";

it("generates some terrain", () => {
    const terrain = new TerrainGenerator(new TerrainSettings());

    const step = 0.1;
    const microsteps = 30;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) / microsteps * step).map(v => terrain.getTerrain(0, v).visualCategory)
    expect(noise).toMatchSnapshot();
});

it("has consistent altitude", () => {
    const terrain = new TerrainGenerator(new TerrainSettings());

    const step = 0.1;
    const microsteps = 30;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) / microsteps * step).map(v => terrain.getTerrain(0, v).altitude)
    expect(noise).toMatchSnapshot();
});

it("has consistent features", () => {
    const terrain = new TerrainGenerator(new TerrainSettings());

    const step = 0.1;
    const microsteps = 30;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2)  / microsteps * step).map(v => terrain.getTerrain(0, v).feature)
    expect(noise).toMatchSnapshot();
});
