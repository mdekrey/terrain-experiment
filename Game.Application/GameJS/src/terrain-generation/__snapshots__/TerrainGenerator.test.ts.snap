// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates some terrain 1`] = `
Array [
  "SnowyMountain",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
  "DeepWater",
]
`;

exports[`has consistent features 1`] = `
Array [
  0.9999999999999998,
  0.5692486381416002,
  0,
  0,
  0.9999999999999998,
  0.9999999999999998,
  0.31573353899508816,
  0,
  0.9999999999999998,
  0.1482383166817639,
  0,
  0,
  0,
  0.48252593447633446,
  0,
  0.0974065349855921,
  0.9999999999999998,
  0.9999999999999998,
  0.30107490828317385,
  0.0890834773195191,
  0.6007595903448548,
  0.9999999999999998,
  0.5632084572883175,
  0.3692563484230739,
  0,
  0.9999999999999998,
  0.12452331880294365,
  0.9999999999999998,
  0.059478670218107266,
  0.9999999999999998,
]
`;
